### PR TITLE
Implement Taproot Sighash

### PR DESCRIPTION
**Pull request by sanket1729**

### CODE REVIEW

**Comment by sanket1729** [comment]
The test seems flaky

**Review by sanket1729 — COMMENTED** [review]
For reference failed logs: https://cirrus-ci.com/task/6028890096271360?logs=ci#L3628

**Comment by sanket1729** [comment]
No diff between `4a7685a` `8202d39`.

File: src/script/interpreter.cpp [hunk 1]
```diff
@@ -1803,6 +1840,40 @@ uint256 GetOutputsSHA256(const T& txTo)
     return ss.GetSHA256();
 }
 
+// Compute a (single) SHA256 of the concatnation of all outputs
+// with padded 0's to internal fields of nValue, nAsset and nNonce.
```
Comment (apoelstra) [inline src/script/interpreter.cpp:1844]: spelling nit: concatenation

File: src/script/interpreter.cpp [hunk 2]
```diff
@@ -1803,6 +1840,40 @@ uint256 GetOutputsSHA256(const T& txTo)
...
+uint256 GetTapOutputsSHA256(const T& txTo)
+{
+    CHashWriter ss(SER_GETHASH, 0);
+    for (const auto& txout : txTo.vout) {
+        txout.SerializePad(ss);
+    }
+    return ss.GetSHA256();
+}
+
+/** Compute the (single) SHA256 of the concatenation of all asset amounts spent by a tx. */
```
Comment (apoelstra) [inline src/script/interpreter.cpp:1855]: "asset amounts" should be "amounts and asset commitments" (similarly for "nonces amounts" below)

Comment (sanket1729) [reply src/script/interpreter.cpp]: Changed to "asset and amount commitments" to match the order of serialization

**Review by sanket1729 — COMMENTED** [review]
Fixed the comments

File: src/script/interpreter.cpp [hunk 3]
```diff
@@ -1954,37 +2035,68 @@ bool SignatureHashSchnorr(uint256& hash_out, const ScriptExecutionData& execdata
...
     assert(execdata.m_annex_init);
     const bool have_annex = execdata.m_annex_present;
     const uint8_t spend_type = (ext_flag << 1) + (have_annex ? 1 : 0); // The low bit indicates whether an annex is present.
     ss << spend_type;
     if (input_type == SIGHASH_ANYONECANPAY) {
-        ss << tx_to.vin[in_pos].prevout;
-        ss << cache.m_spent_outputs[in_pos];
-        ss << tx_to.vin[in_pos].nSequence;
+        if (g_con_elementsmode){
+            ss << (unsigned char)(tx_to.vin[in_pos].prevout.n >> 24);
```
Comment (apoelstra) [inline src/script/interpreter.cpp:2067]: See above comment regarding these shifts by 24 bits

File: test/functional/feature_taproot.py [hunk 4]
```diff
@@ -1106,7 +1120,7 @@ def predict_sigops_ratio(n, dummy_size):
             for witv0 in [False, True]:
                 for hashtype in VALID_SIGHASHES_ECDSA + [random.randrange(0x04, 0x80), random.randrange(0x84, 0x100)]:
                     standard = (hashtype in VALID_SIGHASHES_ECDSA) and (compressed or not witv0)
-                    add_spender(spenders, "legacy/pk-wrongkey", hashtype=hashtype, p2sh=p2sh, witv0=witv0, standard=standard, script=CScript([pubkey1, OP_CHECKSIG]), **SINGLE_SIG, key=eckey1, failure={"key": eckey2}, sigops_weight=4-3*witv0, **ERR_NO_SUCCESS)
+                    add_spender(spenders, "legacy/pk-wrongkey comp=%s wit=%s p2sh=%s hashtype=%s" % (compressed, witv0, p2sh, hashtype), hashtype=hashtype, p2sh=p2sh, witv0=witv0, standard=standard, script=CScript([pubkey1, OP_CHECKSIG]), **SINGLE_SIG, key=eckey1, failure={"key": eckey2}, sigops_weight=4-3*witv0, **ERR_NO_SUCCESS)
```
Comment (apoelstra) [inline test/functional/feature_taproot.py:1123]: did you mean to commit this?

File: test/functional/feature_taproot.py [hunk 5]
```diff
@@ -1303,15 +1316,18 @@ def test_spenders(self, node, spenders, input_counts):
             cur_progress = done / len(spenders)
             next_progress = (done + count_this_tx) / len(spenders)
             change_goal = (1.0 - 0.6 * next_progress) / (1.0 - 0.6 * cur_progress) * balance
-            self.log.debug("Create %i UTXOs in a transaction spending %i inputs worth %.8f (sending ~%.8f to change)" % (count_this_tx, len(unspents), balance * 0.00000001, change_goal * 0.00000001))
+            self.log.info("Create %i UTXOs in a transaction spending %i inputs worth %.8f (sending ~%.8f to change)" % (count_this_tx, len(unspents), balance * 0.00000001, change_goal * 0.00000001))
```
Comment (apoelstra) [inline test/functional/feature_taproot.py:1319]: did you mean to commit this?

File: test/functional/feature_taproot.py (hunk 4 above)
Comment (sanket1729) [reply test/functional/feature_taproot.py:1123]: It was a debug line. Removing it

File: test/functional/feature_taproot.py (hunk 5 above)
Comment (sanket1729) [reply test/functional/feature_taproot.py:1319]: oops

File: src/script/interpreter.cpp [hunk 6]
```diff
@@ -1757,6 +1757,17 @@ class CTransactionSignatureSerializer
     }
 };
 
+/** Compute the (single) SHA256 of the concatenation of all outpoint flags of a tx. */
+template <class T>
+uint256 GetOutpointFlagsSHA256(const T& txTo)
+{
+    CHashWriter ss(SER_GETHASH, 0);
+    for (const auto& txin : txTo.vin) {
+        ss << (unsigned char)(txin.prevout.n >> 24);
```
[the remark this answers is not in the document]
Comment (sanket1729) [reply src/script/interpreter.cpp:1766]: I think what you mean is `((!txin.assetIssuance.IsNull() << 7) + (txin.m_is_pegin << 6))`?

[the remark this answers is not in the document]
Comment (apoelstra) [reply src/script/interpreter.cpp]: Ah yep! Sorry, I was shifting 30 bits in my head rather than 24.

**Comment by apoelstra** [comment]
I think, as long as our test suite passes, there is no need to add new `g_con_elementsmode` uses. We should just drop this and always only do Taproot sighashes in Elements mode.

File: test/functional/test_framework/script.py [hunk 7]
```diff
@@ -793,51 +793,63 @@ def test_cscriptnum_encoding(self):
...
     spk = spent_utxos[input_index].scriptPubKey
-    ss = bytes([0, hash_type]) # epoch, hash_type
+    ss = b""
+    ss += ser_uint256(genesis_hash)
+    ss += ser_uint256(genesis_hash)
+    ss += bytes([hash_type]) # hash_type
     ss += struct.pack("<i", txTo.nVersion)
     ss += struct.pack("<I", txTo.nLockTime)
     if in_type != SIGHASH_ANYONECANPAY:
+        ss += sha256(b"".join(struct.pack("B", ((not i.assetIssuance.isNull()) >> 7) + (i.m_is_pegin >> 6)) for i in txTo.vin))
```
Comment (apoelstra) [inline test/functional/test_framework/script.py:809]: These should be `<<` rather than `>>` (same below)

Comment (sanket1729) [reply test/functional/test_framework/script.py:809]: One more reason to write tests :)

**Comment by apoelstra** [comment]
e7b4bbae4b7da17025bd9fd9926b9956a2d61626 looks great! Just needs some more tests and we're good to go

File: test/functional/test_framework/messages.py [hunk 8]
```diff
@@ -405,6 +405,14 @@ def serialize(self):
         r += self.nInflationKeys.serialize()
         return r
 
+    # serialization used in taproot sighash
+    def tap_sighash_serialize(self):
```
Comment (apoelstra) [inline test/functional/test_framework/messages.py:409]: Can you rename this (and change the comment) to indicate that it is specifically for asset issuance data?

File: test/functional/test_taphash_pegins_issuances.py [hunk 9]
```diff
@@ -0,0 +1,192 @@
+#!/usr/bin/env python3
+# Copyright (c) 2014-2016 The Bitcoin Core developers
+# Distributed under the MIT software license, see the accompanying
+# file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+#
+# Test use of assetdir to locally label assets.
+# Test listissuances returns a list of all issuances or specific issuances based on asset hex or asset label.
+#
```
Comment (apoelstra) [inline test/functional/test_taphash_pegins_issuances.py:9]: This looks great! Can you add it to `test_runner.py` so it gets run in CI?

File: test/functional/test_framework/messages.py (hunk 8 above)
Comment (sanket1729) [reply test/functional/test_framework/messages.py:409]: It is inside the class `CAssetIssuance`, so it is implicit that it is for asset issuance?

Comment (sanket1729) [reply test/functional/test_framework/messages.py:409]: edited the name to `taphash_asset_issuance_serialize`

**Review by sanket1729 — COMMENTED** [review]
Changed the function name and added test to test_runner.py

**Review by apoelstra — APPROVED** [review]
ACK 14f93579a52ee727a9241dab42799aa74ed39b40

Ran all tests (including fuzz tests) several times; did some basic mutation testing on the functional tests to check that obscure things were covered; also poked at the extended tests in #1003 which we are unable to merge.

File: src/script/interpreter.cpp [hunk 10]
```diff
@@ -1757,6 +1757,17 @@ class CTransactionSignatureSerializer
     }
 };
 
+/** Compute the (single) SHA256 of the concatenation of all outpoint flags of a tx. */
+template <class T>
+uint256 GetOutpointFlagsSHA256(const T& txTo)
+{
+    CHashWriter ss(SER_GETHASH, 0);
+    for (const auto& txin : txTo.vin) {
+        ss << (unsigned char) ((!txin.assetIssuance.IsNull() << 7) + (txin.m_is_pegin << 6));
```
Comment (roconnor-blockstream) [inline src/script/interpreter.cpp:1766]: Consider instead using `(COutPoint::OUTPOINT_ISSUANCE_FLAG >> 24)` `COutPoint::OUTPOINT_PEGIN_FLAG >> 24)`

I believe `|` is prefered over `+` for flag composition.

File: src/script/interpreter.cpp [hunk 11]
```diff
@@ -1954,37 +2001,56 @@ bool SignatureHashSchnorr(uint256& hash_out, const ScriptExecutionData& execdata
...
+        ss << cache.m_output_witnesses_single_hash;
     }
-
     // Data about the input/prevout being spent
     assert(execdata.m_annex_init);
     const bool have_annex = execdata.m_annex_present;
     const uint8_t spend_type = (ext_flag << 1) + (have_annex ? 1 : 0); // The low bit indicates whether an annex is present.
     ss << spend_type;
     if (input_type == SIGHASH_ANYONECANPAY) {
+        ss << (unsigned char) ((!tx_to.vin[in_pos].assetIssuance.IsNull() << 7) + (tx_to.vin[in_pos].m_is_pegin << 6));
```
Comment (roconnor-blockstream) [inline src/script/interpreter.cpp:2022]: We should probably abstract this flag composition into its own function as it occurs in two places.

File: src/script/interpreter.cpp [hunk 12]
```diff
@@ -1954,37 +2001,56 @@ bool SignatureHashSchnorr(uint256& hash_out, const ScriptExecutionData& execdata
...
         ss << tx_to.vin[in_pos].prevout;
-        ss << cache.m_spent_outputs[in_pos];
+        ss << cache.m_spent_outputs[in_pos].nAsset;
+        ss << cache.m_spent_outputs[in_pos].nValue;
```
Comment (roconnor-blockstream) [inline src/script/interpreter.cpp:2025]: Maybe comment here that we are not using the nonce and maybe even why we aren't?

**Comment by sanket1729** [comment]
@roconnor-blockstream Addressed in #1023

**Pull request merged**